packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.kt

```diff
@@ -0,0 +1,234 @@
+  private var mModuleRegistry: ModuleRegistry? = null
+  private var mActivityProvider: ActivityProvider? = null
+  private var mActivity: Activity? = null
+
+  // Keep this enum in sync with JavaScript
+  enum class DeviceType(val JSValue: Int) {
+    UNKNOWN(0),
+    PHONE(1),
+    TABLET(2),
+    DESKTOP(3),
+    TV(4);
+  }
+
+  override fun getName(): String {
+    return NAME
+  }
+
+  override fun onCreate(moduleRegistry: ModuleRegistry) {
+    mModuleRegistry = moduleRegistry
+    mActivityProvider = moduleRegistry.getModule(ActivityProvider::class.java)
+    mActivity = mActivityProvider?.currentActivity
+  }
+
+  override fun getConstants(): Map<String, Any> {
+    val constants = HashMap<String, Any>()
+
+    constants["isDevice"] = !isRunningOnGenymotion && !isRunningOnStockEmulator
+    constants["brand"] = Build.BRAND
+    constants["manufacturer"] = Build.MANUFACTURER
+    constants["modelName"] = Build.MODEL
+    constants["designName"] = Build.DEVICE
+    constants["productName"] = Build.PRODUCT
+    constants["deviceYearClass"] = deviceYearClass
+
+    val activityManager = mContext.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
+    val memoryInfo = ActivityManager.MemoryInfo()
+    activityManager.getMemoryInfo(memoryInfo)
+
+    constants["totalMemory"] = memoryInfo.totalMem
+
+    var supportedAbis = Build.SUPPORTED_ABIS
+    if (supportedAbis != null && supportedAbis.isEmpty()) {
+      supportedAbis = null
+    }
+
+    constants["supportedCpuArchitectures"] = supportedAbis
+    constants["osName"] = systemName
+    constants["osVersion"] = Build.VERSION.RELEASE
+    constants["osBuildId"] = Build.DISPLAY
+    constants["osInternalBuildId"] = Build.ID
+    constants["osBuildFingerprint"] = Build.FINGERPRINT
+    constants["platformApiLevel"] = Build.VERSION.SDK_INT
+    constants["deviceName"] = Settings.Secure.getString(mContext.contentResolver, "bluetooth_name")
+
+    return constants
+  }
+
+  private val deviceYearClass: Int
+    get() = YearClass.get(mContext)
+  private val systemName: String
+    get() {
+      var systemName: String
+      if (Build.VERSION.SDK_INT < 23) {
+        systemName = "Android"
+      } else {
+        systemName = Build.VERSION.BASE_OS
+        if (systemName.isEmpty()) {
+          systemName = "Android"
+        }
+      }
+      return systemName
+    }
+
+  @ExpoMethod
+  fun getDeviceTypeAsync(promise: Promise) {
+    val mDeviceType = getDeviceType(mContext)
+    promise.resolve(mDeviceType.JSValue)
+  }
+
+  @ExpoMethod
+  fun getUptimeAsync(promise: Promise) {
+    val uptime = SystemClock.uptimeMillis()
+    promise.resolve(uptime.toDouble())
+  }
+
+  @ExpoMethod
+  fun getMaxMemoryAsync(promise: Promise) {
+    val maxMemory = Runtime.getRuntime().maxMemory()
+    if (maxMemory == Long.MAX_VALUE) {
+      // convert into maximum integer that JS could fit
+      promise.resolve(-1)
+    } else {
+      promise.resolve(maxMemory.toDouble())
+    }
```

Comment:
```suggestion
	// convert into maximum integer that JS could fit
    val maxMemory = Runtime.getRuntime().maxMemory()?.takeIf { it != Long.MAX_VALUE} ?: -1
    promise.resolve(maxMemory.toDouble())
```

Reply:
I've decided to keep semantics exactly as it was ==> integer type is returned when 
`maxMemory == Long.MAX_VALUE`

Therefore:

```kotlin
  @ExpoMethod
  fun getMaxMemoryAsync(promise: Promise) {
    val maxMemory = Runtime.getRuntime().maxMemory()
    promise.resolve(if (maxMemory != Long.MAX_VALUE) maxMemory.toDouble() else -1)
  }
```

It is more concise than initial version && I believe is still readable.